File: cane-csharp/cane-core/Runtime/RuntimeCapabilities.cs
using System;
using System.Collections.Generic;

namespace Cane
{
    public readonly struct RuntimeVersionRange
    {
        public uint MinimumMajor { get; }
        public uint MinimumMinor { get; }
        public uint MaximumMajor { get; }
        public uint MaximumMinor { get; }
        internal RuntimeVersionRange(uint minimumMajor, uint minimumMinor, uint maximumMajor, uint maximumMinor)
        { MinimumMajor = minimumMajor; MinimumMinor = minimumMinor; MaximumMajor = maximumMajor; MaximumMinor = maximumMinor; }
        public bool Contains(uint major, uint minor) =>
            (major > MinimumMajor || (major == MinimumMajor && minor >= MinimumMinor)) &&
            (major < MaximumMajor || (major == MaximumMajor && minor <= MaximumMinor));
    }

    public sealed class RuntimeConformanceInfo
    {
        public string SuiteVersion => "1.8.0";
        public string ManifestSha256 => "ddcd7bb2b5e9dadc00a95d8997886e87c31bc4e5eb5d43811eeabbb2ce98b696";
        public int RequiredCaseCount => 95;
        internal RuntimeConformanceInfo() { }
    }

    /// <summary>Immutable Core metadata. This conformance result does not certify an engine or complete SDK.</summary>
    public sealed class RuntimeCapabilities
    {
        private readonly string[] features = {
            "attachment.bounding-box", "attachment.clipping", "attachment.mesh", "attachment.path", "attachment.point",
            "attachment.region", "attachment.sequence", "blend.add", "blend.multiply", "blend.screen", "constraint.ik",
            "constraint.path", "constraint.physics", "constraint.slider", "constraint.transform", "event", "mesh.deform",
            "mesh.weighted", "skin", "timeline.draw-order", "timeline.skin", "tint.two-color"
        };
        public static RuntimeCapabilities Current { get; } = new RuntimeCapabilities();
        public string ImplementationName => "cane-csharp-runtime";
        public string ImplementationVersion => "0.1.0-alpha.5";
        public string NumericPrecision => "binary32";
        public RuntimeVersionRange RuntimeJson { get; } = new RuntimeVersionRange(1, 0, 1, 0);
        public RuntimeVersionRange Caneb { get; } = new RuntimeVersionRange(1, 0, 1, 0);
        public RuntimeVersionRange Atlas { get; } = new RuntimeVersionRange(1, 0, 1, 0);
        public RuntimeVersionRange RuntimeApi { get; } = new RuntimeVersionRange(1, 0, 1, 3);
        public IReadOnlyList<string> SupportedRuntimeFeatures { get; }
        public ulong ApiFeatureBits => (1UL << 42) - 1;
        public RuntimeConformanceInfo LastPassedConformance { get; } = new RuntimeConformanceInfo();
        private RuntimeCapabilities() { SupportedRuntimeFeatures = Array.AsReadOnly(features); }
        public bool SupportsRuntimeFeature(string? name) => name != null && Array.BinarySearch(features, name, StringComparer.Ordinal) >= 0;
    }
}

File: cane-cpp/cane-core/src/runtime_capabilities.cpp
#include "cane/runtime_capabilities.hpp"
#include <algorithm>
#include <array>

namespace cane {
namespace {
constexpr std::array<std::string_view, 22> features{{
    "attachment.bounding-box", "attachment.clipping", "attachment.mesh", "attachment.path", "attachment.point",
    "attachment.region", "attachment.sequence", "blend.add", "blend.multiply", "blend.screen", "constraint.ik",
    "constraint.path", "constraint.physics", "constraint.slider", "constraint.transform", "event", "mesh.deform",
    "mesh.weighted", "skin", "timeline.draw-order", "timeline.skin", "tint.two-color"
}};
static_assert([] {
    for (std::size_t i = 1; i < features.size(); ++i) if (!(features[i - 1] < features[i])) return false;
    return true;
}(), "Supported feature names must be sorted and unique.");
}
bool RuntimeVersionRange::contains(std::uint32_t major, std::uint32_t minor) const noexcept {
    return (major > minimum.major || (major == minimum.major && minor >= minimum.minor))
        && (major < maximum.major || (major == maximum.major && minor <= maximum.minor));
}
bool supports_runtime_feature(std::string_view name) noexcept { return std::binary_search(features.begin(), features.end(), name); }
RuntimeCapabilities runtime_capabilities() {
    RuntimeCapabilities result;
    result.implementation_name = "cane-cpp-runtime"; result.implementation_version = "0.1.0"; result.numeric_precision = "binary32";
    result.runtime_json = {{1, 0}, {1, 0}}; result.caneb = result.runtime_json; result.atlas = result.runtime_json;
    result.runtime_api = {{1, 0}, {1, 3}};
    result.supported_runtime_features.assign(features.begin(), features.end());
    result.api_feature_bits = (std::uint64_t{1} << 42) - 1;
    result.last_passed_conformance = {"1.8.0", "ddcd7bb2b5e9dadc00a95d8997886e87c31bc4e5eb5d43811eeabbb2ce98b696", 95};
    return result;
}
}
